### PR TITLE
Retry with backoff on cluster connection failures (II)

### PR DESCRIPTION
**Pull request by sazzad16**
Original PR #2358 (by @walles)

### CODE REVIEW

File: src/main/java/redis/clients/jedis/JedisClusterCommand.java [hunk 1]
```diff
@@ -132,14 +159,60 @@ private T runWithRetries(final int slot) {
     throw maxAttemptsException;
```
Comment (yangbodong22011) [inline src/main/java/redis/clients/jedis/JedisClusterCommand.java:159]: If `maxAttempts = 0`, we will get following exception():
```
Exception in thread "main" java.lang.NullPointerException: Cannot suppress a null exception.
	at java.base/java.lang.Throwable.addSuppressed(Throwable.java:1057)
	at redis.clients.jedis.JedisClusterCommand.runWithRetries(JedisClusterCommand.java:158)
	at redis.clients.jedis.JedisClusterCommand.run(JedisClusterCommand.java:45)
	at redis.clients.jedis.JedisCluster.set(JedisCluster.java:292)
	at redis.clients.jedis.tests.FancheTest.main(FancheTest.java:27)
```

but in the old code, we will get ` No more cluster attempts left.`, should we define default `lastException = maxAttemptsException`?

Comment (sazzad16) [reply src/main/java/redis/clients/jedis/JedisClusterCommand.java:159]: @yangbodong22011 
1. This is the current behavior at `master` branch.
2. Why would you send `maxAttempts = 0`?

Comment (yangbodong22011) [reply src/main/java/redis/clients/jedis/JedisClusterCommand.java:159]: > @yangbodong22011
> 
> 1. This is the current behavior at `master` branch.
> 2. Why would you send `maxAttempts = 0`?

1. ok, ignore that exception
2. just a test, or we should disable user Settings `maxAttempts = 0` (params check may be another PR), this PR can continue.

**Review by yangbodong22011 — APPROVED** [review]

File: src/main/java/redis/clients/jedis/JedisClusterCommand.java [hunk 2]
```diff
@@ -85,7 +100,10 @@ public T runWithAnyNode() {
   }
 
   private T runWithRetries(final int slot) {
+    Instant deadline = Instant.now().plus(maxTotalRetriesDuration);
```
Comment (gkorland) [inline src/main/java/redis/clients/jedis/JedisClusterCommand.java:103]: It feels like we adding performance impact with calling now() on each command run.

**Review by gkorland — APPROVED** [review]

**Comment by sazzad16** [comment]
Everything pushed to the original PR #2358

**Pull request closed without merging**